File: NotesStudy/OC/Commons/HomePageMenu/HomePageMenuView.m
//
//  HomePageMenuView.m
//  NoteDome
//
//  Created by Lj on 2018/1/18.
//  Copyright © 2018年 Lj. All rights reserved.
//

#import "HomePageMenuView.h"
#import "HomePageMenuCell.h"

static CGFloat const CorrectNumber = 0.2;

static NSString *homePageMenuCell = @"HomePageMenuCell";

@interface HomePageMenuView () <UICollectionViewDelegate,UICollectionViewDataSource,UICollectionViewDelegateFlowLayout>
@property (nonatomic, strong) NSArray *nameArray;
@property (nonatomic, strong) NSArray *iconArray;
@property (nonatomic, strong) NSArray *numberArray;
@property (nonatomic, strong) UICollectionViewFlowLayout *flowLayout;

@end

@implementation HomePageMenuView

- (id)initWithFrame:(CGRect)frame {
    self.flowLayout = [[UICollectionViewFlowLayout alloc]init];
    [self configDefaultValues];
    NSLog(@"------%ld------",self.column);
    if (self = [self initWithFrame:frame collectionViewLayout:self.flowLayout]) {
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame collectionViewLayout:(UICollectionViewLayout *)layout {
    if (self = [super initWithFrame:frame collectionViewLayout:layout]) {
        self.backgroundColor = [UIColor whiteColor];
        self.flowLayout = (UICollectionViewFlowLayout *)layout;
        self.collectionViewLayout = self.flowLayout;
        NSLog(@"------%ld------",self.column);
        [self setUpUI];
    }
    return self;
}

- (void)setUpUI {
    self.delegate = self;
    self.dataSource = self;
    self.alwaysBounceVertical = NO;
    if (@available(iOS 11.0, *)) {
        self.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    } else {
        // Fallback on earlier versions
    }
    [self registerClass:[HomePageMenuCell class] forCellWithReuseIdentifier:homePageMenuCell];    
}

- (void)configDefaultValues {
    self.interitemSpacing = 1;
    self.lineSpacing = 1;
    _column = 4;
    _iconImageSize = 50;
    _itemSizeHeight = 90;
}


- (void)reloadData {
    [super reloadData];
    
    CGFloat allWidth = CGRectGetWidth(self.frame) - (_column - 1) * _interitemSpacing - CorrectNumber;
    self.itemSize = CGSizeMake(allWidth/_column, _itemSizeHeight);
    self.nameArray = [self.menuDataSource homePageMenuName:self];
    CGRect frame = self.frame;
    if (self.nameArray.count == 0) frame.size.height = 0;
    else {
        frame.size.height = (_itemSize.height + _lineSpacing) * (self.nameArray.count/_column + (self.nameArray.count%_column == 0 ? 0 : 1)) + CorrectNumber - _lineSpacing;
    }
    self.frame = frame;
    if (self.menuDataSource && [self.menuDataSource respondsToSelector:@selector(homePageMenuImage:)]) {
        self.iconArray = [self.menuDataSource homePageMenuImage:self];
    }
    if (self.menuDataSource && [self.menuDataSource respondsToSelector:@selector(homePageMenuNumber:)]) {
        self.numberArray = [self.menuDataSource homePageMenuNumber:self];
    }
}

- (void)dealloc {
    NSLog(@"------dealloc------");
}

- (void)setColumn:(NSInteger)column {
    _column = column;
}

- (void)setLineSpacing:(CGFloat)lineSpacing {
    _lineSpacing = lineSpacing;
    self.flowLayout.minimumLineSpacing = _lineSpacing;
}

- (void)setInteritemSpacing:(CGFloat)interitemSpacing {
    _interitemSpacing = interitemSpacing;
    self.flowLayout.minimumInteritemSpacing = _interitemSpacing;
}

- (void)setItemSize:(CGSize)itemSize {
    _itemSize = itemSize;
}

#pragma mark - UICollectionViewDelegate,UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.nameArray.count;
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    HomePageMenuCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:homePageMenuCell forIndexPath:indexPath];
    if (cell == nil) {
        cell = [[HomePageMenuCell alloc]init];
    }
    [self _collectionView:cell cellForItemAtIndexPath:indexPath];
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    if (self.menuDelegate && [self.menuDelegate respondsToSelector:@selector(menuView:didSelectItemAtIndexPath:)]) {
        [self.menuDelegate menuView:self didSelectItemAtIndexPath:indexPath];
    }
    if (self.menuDelegate && [self.menuDelegate respondsToSelector:@selector(menuView:didSleectCell:itemAtIndexPath:)]) {
        HomePageMenuCell *cell = (HomePageMenuCell *)[collectionView cellForItemAtIndexPath:indexPath];
        [self.menuDelegate menuView:self didSleectCell:cell itemAtIndexPath:indexPath];
    }
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    if (self.menuDataSource && [self.menuDataSource respondsToSelector:@selector(homePageMenuView:sizeForItemAtIndexPath:)]) {
        return [self.menuDataSource homePageMenuView:self sizeForItemAtIndexPath:indexPath];
    }else {
        return self.itemSize;
    }
}

#pragma mark -
- (void)_collectionView:(HomePageMenuCell *)cell cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.item == 0) cell.backgroundColor = [UIColor redColor];
    else cell.backgroundColor = [UIColor blueColor];
    
    cell.iconImage.backgroundColor = [UIColor whiteColor];
    cell.iconImageSize = self.iconImageSize;
    cell.nameLabel.text = self.nameArray[indexPath.item];
    if (self.iconArray.count != 0) {
        NSCAssert(self.nameArray.count == self.iconArray.count, @"The icon is different from the number of names");
        cell.iconImage.image = [UIImage imageNamed:self.iconArray[indexPath.item]];
    }
    if (self.numberArray.count == self.nameArray.count) {
        cell.numberStr = self.numberArray[indexPath.item];
    }else cell.numberLabel.hidden = YES;
}




@end

File: SwiftNoteStudy/SwiftNoteStudy/Commons/HomePageMenu/HomePageMenuView.swift
//
//  HomePageMenuView.swift
//  SwiftNoteStudy
//
//  Created by less on 2018/11/2.
//  Copyright © 2018 less. All rights reserved.
//

import UIKit
import Kingfisher

protocol HomePageMenuDelegate: NSObjectProtocol {
    /** item点击 */
    func menuView(_ menuView: HomePageMenuView, didSelectItemAt indexPath: IndexPath)
    /** item点击 */
    func menuView(_ menuView: HomePageMenuView, collectionCell: UICollectionViewCell,didSelectItemAt indexPath: IndexPath)
    /** View高度 */
    func menuViewHeight(_ menuView: HomePageMenuView, viewHeight: CGFloat)
    /** 3D touch */
    func menuView(_ menuView: HomePageMenuView, previewingContext: UIViewControllerPreviewing, touchItemAt indexPath: IndexPath) -> UIViewController?
}

protocol HomePageMenuDataSource: NSObjectProtocol {
    /** 名字 */
    func menuViewName(_ menuView: HomePageMenuView) -> [String]
    /** 图片 */
    func menuViewImage(_ menuView: HomePageMenuView) -> [String]
    
    /** 消息条数 */
    func menuViewNumber(_ menuView: HomePageMenuView) -> [Int]
    /** cell大小 */
    func menuViewSize(_ menuView: HomePageMenuView, forItemAt indexPath: IndexPath) -> CGSize

}

extension HomePageMenuDelegate {
    func menuView(_ menuView: HomePageMenuView, didSelectItemAt indexPath: IndexPath) {}
    func menuView(_ menuView: HomePageMenuView, collectionCell: UICollectionViewCell,didSelectItemAt indexPath: IndexPath) {}
    func menuViewHeight(_ menuView: HomePageMenuView, viewHeight: CGFloat) {}
    func menuView(_ menuView: HomePageMenuView, previewingContext: UIViewControllerPreviewing, touchItemAt indexPath: IndexPath) -> UIViewController? {
        return nil
    }
}

extension HomePageMenuDataSource {
    func menuViewNumber(_ menuView: HomePageMenuView) -> [Int] {
        return [0]
    }
    
    func menuViewSize(_ menuView: HomePageMenuView, forItemAt indexPath: IndexPath) -> CGSize {
        return CGSize.zero
    }
    
}


class HomePageMenuView: UICollectionView {
    weak var menuDelegate: HomePageMenuDelegate?
    weak var menuDataSource: HomePageMenuDataSource?
    /** 横向数量 默认itemSize 有效 */
    public var column: Int = 4
    /** view最大高度 isAdaptiveHeight = true 生效 */
    public var memuHeight: CGFloat = (Constant.screenHeight - Constant.bottomBarHeight - Constant.navbarAndStatusBar - 50)
    
    /** 是否自适应高度 */
    public var isAdaptiveHeight: Bool = false
    
    /** 是否支持3D touch */
    public var is3DTouch: Bool = false
    
    /** 横向间隔 */
    public var interitemSpacing: CGFloat = 1 {
        didSet {
            flowLayout.minimumInteritemSpacing = interitemSpacing
        }
    }
    /** 纵向间距 */
    public var lineSpacing: CGFloat = 1 {
        didSet {
            flowLayout.minimumLineSpacing = lineSpacing
        }
    }
    /** 周边间隔 */
    public var sectionInset: UIEdgeInsets = UIEdgeInsets() {
        didSet {
            flowLayout.sectionInset = sectionInset
        }
    }
    /** cell属性 */
    public var cellStruct: MenuCellStruct = MenuCellStruct()
    /** itemHeight */
    public var itemSizeHeight: CGFloat = 70
    
    fileprivate let CorrectNumber: CGFloat = 0.2
    fileprivate var flowLayout: UICollectionViewFlowLayout!
    fileprivate var itemSize: CGSize = CGSize.zero
    fileprivate var nameArray = [String]()
    fileprivate var iconArray = [String]()
    fileprivate var numberArray = [Int]()

    deinit {
        print("----deinit------")
    }
    
    convenience init(frame: CGRect) {
        self.init(frame: frame, collectionViewLayout: UICollectionViewFlowLayout())
    }
    
    override init(frame: CGRect, collectionViewLayout layout: UICollectionViewLayout) {
        super.init(frame: frame, collectionViewLayout: layout)
        backgroundColor = UIColor.clear
        flowLayout = layout as? UICollectionViewFlowLayout
        collectionViewLayout = layout
        initView()
    }
    
    fileprivate func initView() {
        self.delegate = self
        self.dataSource = self
        self.alwaysBounceVertical = false
        flowLayout.minimumInteritemSpacing = interitemSpacing
        flowLayout.minimumLineSpacing = lineSpacing

        if #available(iOS 11.0, *) {
            self.contentInsetAdjustmentBehavior = .never
        } else {
        }
        self.register(HomePageMenuCell.self, forCellWithReuseIdentifier: HomePageMenuCell.identifire)
    }
    
    override func reloadData() {
        super.reloadData()
        
        let allWidth: CGFloat = frame.width - CGFloat((column - 1)) * interitemSpacing - CorrectNumber - sectionInset.left - sectionInset.right
        self.itemSize = CGSize(width: allWidth/CGFloat(column), height: itemSizeHeight)
        if let menuData = menuDataSource {
            nameArray = menuData.menuViewName(self)
            iconArray = menuData.menuViewImage(self)
            numberArray = menuData.menuViewNumber(self)
        }
        menuDelegate?.menuViewHeight(self, viewHeight: flowLayout.collectionViewContentSize.height)
        if nameArray.count == 0 { frame.size.height = 0 }
        assert(nameArray.count == iconArray.count, "The name is not equal to the number of images")
        if isAdaptiveHeight {
            self.height = min(memuHeight, flowLayout.collectionViewContentSize.height)
        }
    }
    
    fileprivate func collectionCell(_ cell: HomePageMenuCell, cellForItemAt indexPath: IndexPath) {
        cell.nameLabel.text = nameArray[indexPath.item]
        cell.cellStruct = cellStruct
        if nameArray.count == numberArray.count {
            cell.number = numberArray[indexPath.item]
        }
        let imageStr = iconArray[indexPath.item]
        if imageStr.hasPrefix("http://") {
            if let url = URL(string: imageStr) {
                cell.iconImage.kf.setImage(with: ImageResource(downloadURL: url), placeholder: UIImage(named: "ddd"))
            }
        }else {
            cell.iconImage.image = UIImage(named: imageStr)
        }
        
    }
    
    fileprivate func registerForPreviewingView(_ cell: HomePageMenuCell, cellForItemAt indexPath: IndexPath) {
        if traitCollection.forceTouchCapability == .available && is3DTouch {
            // 支持3D Touch
            // 注册Peek & Pop功能
            self.viewController()?.registerForPreviewing(with: self, sourceView: cell)
        }
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

extension HomePageMenuView: UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return nameArray.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: HomePageMenuCell.identifire, for: indexPath) as! HomePageMenuCell
        collectionCell(cell, cellForItemAt: indexPath)
        registerForPreviewingView(cell, cellForItemAt: indexPath)
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        let indexSize = menuDataSource?.menuViewSize(self, forItemAt: indexPath)
        if indexSize!.equalTo(CGSize.zero) {
            return itemSize
        }else {
            return indexSize!
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        menuDelegate?.menuView(self, didSelectItemAt: indexPath)
        let cell = collectionView.cellForItem(at: indexPath)
        menuDelegate?.menuView(self, collectionCell: cell!, didSelectItemAt: indexPath)
    }
    
}

extension HomePageMenuView: UIViewControllerPreviewingDelegate {
    func previewingContext(_ previewingContext: UIViewControllerPreviewing, viewControllerForLocation location: CGPoint) -> UIViewController? {
        let indexPath = self.indexPath(for: previewingContext.sourceView as! UICollectionViewCell)
        if let index = indexPath {
            return menuDelegate?.menuView(self, previewingContext: previewingContext, touchItemAt: index)
        }
        return nil
    }
    
    func previewingContext(_ previewingContext: UIViewControllerPreviewing, commit viewControllerToCommit: UIViewController) {
        self.viewController()?.show(viewControllerToCommit, sender: self)
    }
}

struct MenuCellStruct {
    /** 图片距上距离 */
    var iconImageTop: CGFloat = 5
    /** 图标宽高 */
    var iconImageHeight: CGFloat = 40
    /** 文字距图片距离 */
    var titleTopHeight: CGFloat = 5
    /** 文字大小 */
    var titleFont: UIFont = UIFont.systemFont(ofSize: 14)
    /** 文字颜色 */
    var titleColor: UIColor = UIColor.black
}
